File: simulate/results/plot.py
import pandas as pd
import sqlite3
import matplotlib.pyplot as plt

η_c_values = [1.0, 0.9, 0.5, 0.3]
ϵ_g_values = [0.0001, 0.001]

conn = sqlite3.connect("./simulate/results/results3.db")
df = pd.read_sql_query("SELECT * FROM \"QNet-MTP\"", conn).sort_values(by=["ϵ_g", "η_c", "n", "L"])

print(df)

fig, axes = plt.subplots(2, 4, sharex=True, sharey=True, figsize=(20, 10))

for i, ϵ_g in enumerate(ϵ_g_values):
    for j, η_c in enumerate(η_c_values):
        ax = axes[i, j]
        filtered_df = df[(df["ϵ_g"] == ϵ_g) & (df["η_c"] == η_c)]
        grouped_df = filtered_df.groupby("n")

        for name, group in grouped_df:
            ax.plot(group["L"], group["E_Y"], label=f"n={name}")

        ax.set_xscale("log", base=10)
        ax.set_yscale("log", base=10)
        ax.set_title(f"ϵ_g={ϵ_g}, η_c={η_c}")
        ax.legend()

plt.tight_layout(rect=[0.03, 0.03, 1, 0.95])

plt.savefig("./simulate/results/results3.png")
print("File saved.")

File: simulate/results/plot.jl
using Plots
using LaTeXStrings
using Plots.PlotMeasures

using ColorSchemes
palette = ColorSchemes.mk_12

using SQLite, DataFrames
db = SQLite.DB("./simulate/results/results4.db")
df = DBInterface.execute(db, "SELECT * FROM \"QNet-MTP\"") |> DataFrame
sort!(df, [:n, :η_c, :ϵ_g, :L])

η_c_values = [1.0, 0.9, 0.5, 0.3]
ϵ_g_values = [0.0001, 0.001]


plot_layout = @layout [a b c d; e f g h]
p = plot(layout = plot_layout, size=(2000, 1000))

xlabel!(p, "Distance (km)")
ylabel!(p, "Secret key rate \n (bits per channel use per burst)")

for (i, ϵ_g) in enumerate(ϵ_g_values)
    for (j, η_c) in enumerate(η_c_values)
        filtered_df = filter(row -> row.ϵ_g == ϵ_g && row.η_c == η_c, df)
        grouped_df = groupby(filtered_df, :n)

        for g in grouped_df
            n = g[1, :n]; iter = convert(Int, log2(n)) + 1
            distance_vec = g[!, :L] ./ 1000
            real_vec = g[!, :SKR]

            plot!(p[i, j], distance_vec, real_vec,
                title=L"ϵ_g = %$(ϵ_g), η_c = %$(η_c)", 
                legend_font_pointsize=10, 
                legend_title_font=10, 
                label="N = $n", 
                lw=3.5, 
                xlims=(1e1, 1e4),
                ylim=(1e-4, 1e0), 
                color=palette[iter],
                dpi=500, 
                left_margin = 5mm ,
                right_margin = 5mm,
                bottom_margin = 5mm, 
                top_margin = 4mm,
                legend = :none
            )
        end
    end
end

plot!(p, xscale=:log10, yscale=:log10,)
savefig(p, "simulate/results/results4.png")
